TASK: refactor nodes loading

dimaip opened the pull request:
- Eases the pain of slow nodes state rendering by fetching default nodes asynchronously (#1579)
- Unify logic for re-fetching nodes state
- Reload nodes when switching workspaces, redirect to correct node if it doesn't exist in target (Fixes: #1464)
- Reloads the tree in one gulp, without extra requests
- Fixes loading indicator
- Loads default nodes + toggled nodes + nodes up to the current document node in one request
- Parallelizes JSON fetch requests for translations and other stuff (Improves #1020)
- Don't render hidden elements (improves render time from 1100ms down to 120-360ms!) (Fixes: #887 )

daniellienert commented on Classes/Neos/Neos/Ui/Controller/BackendServiceController.php:
```diff
@@ -237,10 +245,11 @@ public function discardAction(array $nodeContextPaths)
      * @return void
      */
-    public function changeBaseWorkspaceAction($targetWorkspaceName)
+    public function changeBaseWorkspaceAction($targetWorkspaceName, $documentNode)
```
Parameters should be annotated `string $targetWorkspaceName, NodeInterface $documentNode`

daniellienert commented on Classes/Neos/Neos/Ui/Controller/BackendServiceController.php:
```diff
@@ -262,8 +271,35 @@ public function changeBaseWorkspaceAction($targetWorkspaceName)
+                } else {
+                    $redirectNode = $redirectNode->getParent();
+                    if (!$redirectNode) {
+                        throw new \Exception('Wasn\'t able to locate any valid node in rootline in the target workspace.');
```
Exception should have an exception number

daniellienert replied:
Add the documentnodes identifier to the exception to support debugging

daniellienert commented on Classes/Neos/Neos/Ui/Domain/Model/Feedback/Operations/Redirect.php:
```diff
@@ -0,0 +1,101 @@
+<?php
+namespace Neos\Neos\Ui\Domain\Model\Feedback\Operations;
+
```
Licence header missing

dimaip replied:
It's missing for all files in the package...

daniellienert commented on Classes/Neos/Neos/Ui/Domain/Model/Feedback/Operations/Redirect.php:
```diff
@@ -0,0 +1,101 @@
+     */
+    public function getDescription()
+    {
+        return sprintf('Redirect to node "%s".', $this->getNode()->getContextPath());
```
input type and return type declaration are missing at many placed, would be nice to have them

daniellienert commented on Classes/Neos/Neos/Ui/Domain/Model/Feedback/Operations/Redirect.php:
```diff
@@ -0,0 +1,101 @@
+     * Serialize the payload for this feedback
+     *
+     * @param ControllerContext $controllerContext
+     * @return mixed
```
Array is returned here

daniellienert commented on Classes/Neos/Neos/Ui/FlowQueryOperations/NeosUiDefaultNodesOperation.php:
```diff
@@ -55,28 +55,44 @@ public function canEvaluate($context)
-        list($baseNodeType, $loadingDepth) = array_replace([
+        list($siteNode, $documentNode) = $flowQuery->getContext();
+        // set default values for arguments
+        list($baseNodeType, $loadingDepth, $toggledNodes) = array_replace([
```
That looks risky (or I don't get it). If $arguments contains 3 values the array replace could be skipped. If it contains less, How do you know here which arguments are set?

dimaip replied:
Agreed, it was not my code, took me some time to figure out what it's doing actually. I think defaults are not needed here, will remove it.

dimaip commented:
@daniellienert thanks, fixed your comments.
Did you actually test the change itself? Would be great to get some extended testing on real projects, as I changed a lot of stuff.

skurfuerst commented on Classes/Neos/Neos/Ui/Controller/BackendServiceController.php:
```diff
@@ -262,8 +272,35 @@ public function changeBaseWorkspaceAction($targetWorkspaceName)
+
+            // If current document node doesn't exist in the base workspace, traverse its parents to find the one that exists
+            $redirectNode = $documentNode;
+            while (true) {
```
uhm, I have to check this while(true) logic in detail; it's always hard to see that it does not end up with an endless loop ;)

dimaip replied:
Sure, feel free to refactor if you know how to express it better. But I think it's safe.

skurfuerst commented on Classes/Neos/Neos/Ui/FlowQueryOperations/NeosUiDefaultNodesOperation.php:
```diff
@@ -17,7 +17,7 @@
 
 /**
- * Renders nodes that are initially required by the user interface
+ * Fetches all nodes needed for the given state of the UI
```
isn't this function actually *slower* than the one before? :)

dimaip replied:
For default state, with no uncollapsed nodes, it's equivalent. It includes nodes on the rootline + all uncollapsed nodes so extra requests would not be needed when opening on a deep nested node, deeper then loadingDepth, or when reloading the tree.

skurfuerst commented on packages/neos-ui/src/Containers/Drawer/index.js:
```diff
@@ -113,7 +113,7 @@ export default class Drawer extends PureComponent {
                 aria-hidden={isHidden ? 'true' : 'false'}
                 >
-                {Object.values(menuData).map((item, index) => (
+                {!isHidden && Object.values(menuData).map((item, index) => (
```
really nice :)

skurfuerst commented on packages/neos-ui/src/Sagas/UI/PageTree/index.js:
```diff
@@ -78,26 +78,6 @@ export function * watchNodeCreated() {
-            const node = nodesToReload[i];
-            const contextPath = node.get('contextPath');
-
-            yield put(actions.UI.PageTree.requestChildren(contextPath, {unCollapse: false}));
```
OK. in the old code, we had many AJAX requests, now we only have one.

dimaip replied:
exactly, works a lot faster

skurfuerst commented on packages/neos-ui/src/index.js:
```diff
@@ -176,6 +182,14 @@ function * application() {
+        siteNodeContextPath,
+        documentNodeContextPath,
+        merge: true
+    }));
```
I am wondering if this change actually improves performance; or if it improves *perceived* performance. As long as the nodes are not loaded, can we even interact with the UI at all?

dimaip replied:
It improves perceived performance. You can start interacting with everything, e.g. start inline editing. The nodes are seamlessly added when loaded. It takes 100ms to add 50 nodes to the tree, so it doesn't even lag.

kdambekalns replied:
It's "actual perceived performance" ;) If you read about the timings in https://github.com/neos/neos-ui/issues/1579 you get an idea. So, the perceived performance increases (you don't just see the spinner for ages), and the actual performance as well, because you can actually start doing $something.

dimaip replied:
@kdambekalns see my RFC how to increase real performance: https://github.com/neos/neos-ui/issues/1646

skurfuerst commented:
Following up on your measurements in https://github.com/neos/neos-ui/pull/1581#issuecomment-361018832:

- I think we should remove the PolicyAspect and *directly* inline the code in the NodeInfoHelper.
- is AOP slow or the policy check itself?

all the best,
Sebastian

skurfuerst commented:
btw - ❤️ for working on this - it is great to have that tackled!

dimaip commented:
@skurfuerst hey thanks a lot for the review!

> I think we should remove the PolicyAspect and directly inline the code in the NodeInfoHelper.

That will not help. AOP cost is negligible compared to how much it costs to evaluate EEL for each node. I did xdebug profiling.

I think to solve the problem we should work in multiple directions:

a) make tree loading async, as done in this PR: partially hides the problem, but it's not enough if the editors want to interact with the tree right away
b) separate loading node policies from loading node data, and make it also async. So in first pass the whole UI would be rendered, then with a tiny delay the tree nodes would appear and some time after the policies would kick in.
c) improve performance of actual policy checking and eel parsing. But it's a whole other topic.

dimaip commented:
@skurfuerst so here's what I suggest to do next after we merge this PR: https://github.com/neos/neos-ui/issues/1646

skurfuerst commented:
Awesome; I have a big instance now where I will test today :)

skurfuerst commented:
## Performance measurements (WIP)

I used [speedline](https://github.com/paulirish/speedline) to measure performance. The red line is when editing is possible.


## baseline - old UI - no optimizations

<img width="582" alt="downloads zsh 2018-02-23 10-00-04" src="https://user-images.githubusercontent.com/190777/36586033-728bfade-1880-11e8-8483-dc1c57aa2177.png">

## new UI -- without this patch

<img width="549" alt="downloads zsh 2018-02-23 10-01-28" src="https://user-images.githubusercontent.com/190777/36586055-8ce42122-1880-11e8-9a2f-894ec474b83b.png">

## new UI - with patch (TODO)

skurfuerst commented:
@dimaip I found the following issue:

- as soon as the page finishes loading, the spinner in the node tree stops; although the node tree is not yet fully loaded. I'll push a followup for this.

kdambekalns commented:
> Let's merge as soon as Travis OKs it!

Well… 😏

dimaip commented:
@kdambekalns ?
